ODIN_II/SRC/HardSoftLogicMixer.cpp

```diff
@@ -0,0 +1,237 @@
+/*
+ * Permission is hereby granted, free of charge, to any person
+ * obtaining a copy of this software and associated documentation
+ * files (the "Software"), to deal in the Software without
+ * restriction, including without limitation the rights to use,
+ * copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following
+ * conditions:
+ *
+ * The above copyright notice and this permission notice shall be
+ * included in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
+ * OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+ * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+ * WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ */
+#include "HardSoftLogicMixer.hpp"
+#include <iostream>
+#include <vector>
+#include <cstdint>
+#include <cstring>
+#include "OdinGridAnalyzer.hpp"
+#include <bits/stdc++.h>
+#include "partial_map.h" // instantiate add_with_carry
+#include "adders.h"      // instantiate_hard_adders
+
+HardSoftLogicMixer::HardSoftLogicMixer(t_arch& arch, const config_t config, std::vector<t_physical_tile_type> tileTypes)
+    : _analyzer()
+    , _arch(arch) {
+    _allOptsDisabled = true;
+    _tileTypes = tileTypes;
+    // By default, disables all optimizations
+    for (int i = 0; i < mix_hard_blocks::Count; i++) {
+        _enabledOptimizations[i] = false;
+    }
+
+    parseAndSetOptimizationParameters(config);
+}
+
+void HardSoftLogicMixer::parseAndSetOptimizationParameters(const config_t config) {
+    if (config.mix_soft_and_hard_logic != 0) {
+        int checkValue = -1;
+        int mix_soft_and_hard_logic = config.mix_soft_and_hard_logic;
+        int temp;
+        for (int i = 0; i < mix_hard_blocks::Count; i++) {
+            temp = 1 << (i);
+            checkValue = temp & mix_soft_and_hard_logic;
+            if (checkValue != 0) {
+                _enabledOptimizations[i] = true;
+            }
+        }
+        _allOptsDisabled = false;
+        //Explicitly set all hard block numbers to max
+        for (int i = 0; i < mix_hard_blocks::Count; i++) {
+            _hardBlocksCount[i] = INT_MAX;
+            _hardBlocksMixingRatio[i] = 1.0f;
+        }
+
+        _hardBlocksMixingRatio[mix_hard_blocks::MULTIPLIERS] = config.mults_mixing_ratio;
+
+        if (config.mults_mixing_ratio >= 0.0f && config.mults_mixing_ratio <= 1.0f) {
+            _enabledOptimizations[mix_hard_blocks::MULTIPLIERS] = true;
+            _allOptsDisabled = false;
+        }
+
+        if (config.mults_mixing_exact_number_of_multipliers >= 0) {
```

Comment:
I think those config variable need to be removed, those initializations should happen during the XML parsing, so maybe write this piece with this in mind, we are keeping two variables for no really specific reason